File: zoomx/src/main/java/com/zoomx/zoomx/model/RequestEntity.java
package com.zoomx.zoomx.model;

import android.arch.persistence.room.Entity;
import android.arch.persistence.room.PrimaryKey;
import android.arch.persistence.room.TypeConverters;

import com.zoomx.zoomx.db.converters.HeaderConverter;

import java.util.ArrayList;
import java.util.Date;
import java.util.Map;

/**
 * Created by Ahmed Fathallah on 11/20/2017.
 */

@Entity(tableName = "requests")
public class RequestEntity {
    @PrimaryKey(autoGenerate = true)
    private int id;
    private int code;
    private String method;
    private Date startDate;
    private String url;
    private String responseBody;
    private String requestBody;
    private long tookTime;
    private long responseSizeInBytes;

    @TypeConverters(HeaderConverter.class)
    private HeaderViewModel requestHeaders;
    @TypeConverters(HeaderConverter.class)
    private HeaderViewModel responseHeaders;

    public RequestEntity() {
    }

    private RequestEntity(int code, String method, Date startDate, String url
            , String responseBody, String requestBody, long tookTime, long responseSizeInBytes
            , HeaderViewModel requestHeaders, HeaderViewModel responseHeaders) {
        this.code = code;
        this.method = method;
        this.startDate = startDate;
        this.url = url;
        this.responseBody = responseBody;
        this.requestBody = requestBody;
        this.tookTime = tookTime;
        this.responseSizeInBytes = responseSizeInBytes;
        this.requestHeaders = requestHeaders;
        this.responseHeaders = responseHeaders;
    }

    //region Setter and getter
    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public int getCode() {
        return code;
    }

    public void setCode(int code) {
        this.code = code;
    }

    public String getMethod() {
        return method != null ? method : "";
    }

    public void setMethod(String method) {
        this.method = method;
    }

    public Date getStartDate() {
        return startDate;
    }

    public void setStartDate(Date startDate) {
        this.startDate = startDate;
    }

    public String getUrl() {
        return url != null ? url : "missing";
    }

    public void setUrl(String url) {
        this.url = url;
    }

    public String getResponseBody() {
        return responseBody != null ? responseBody : "";
    }

    public void setResponseBody(String responseBody) {
        this.responseBody = responseBody;
    }

    public String getRequestBody() {
        return requestBody != null ? requestBody : "";
    }

    public void setRequestBody(String requestBody) {
        this.requestBody = requestBody;
    }

    public long getTookTime() {
        return tookTime;
    }

    public void setTookTime(long tookTime) {
        this.tookTime = tookTime;
    }

    public long getResponseSizeInBytes() {
        return responseSizeInBytes;
    }

    public void setResponseSizeInBytes(long responseSizeInBytes) {
        this.responseSizeInBytes = responseSizeInBytes;
    }

    public HeaderViewModel getRequestHeaders() {
        return requestHeaders;
    }

    public void setRequestHeaders(HeaderViewModel requestHeaders) {
        this.requestHeaders = requestHeaders;
    }

    public HeaderViewModel getResponseHeaders() {
        return responseHeaders;
    }

    public void setResponseHeaders(HeaderViewModel responseHeaders) {
        this.responseHeaders = responseHeaders;
    }

    //endregion

    //region builder
    public static class Builder {

        private int code;
        private String method;
        private Date startDate;
        private String url;
        private String responseBody;
        private String requestBody;
        private long tookTime;
        private long responseSizeInBytes;
        private ArrayList<String> requestHeaders;
        private ArrayList<String> responseHeaders;

        public Builder setCode(int code) {
            this.code = code;
            return this;
        }

        public Builder setMethod(String method) {
            this.method = method;
            return this;
        }

        public Builder setStartDate(Date startDate) {
            this.startDate = startDate;
            return this;
        }

        public Builder setUrl(String url) {
            this.url = url;
            return this;
        }

        public Builder setResponseBody(String responseBody) {
            this.responseBody = responseBody;
            return this;
        }

        public Builder setRequestBody(String requestBody) {
            this.requestBody = requestBody;
            return this;
        }

        public Builder setTookTime(long tookTime) {
            this.tookTime = tookTime;
            return this;
        }

        public Builder setResponseSizeInBytes(long responseSizeInBytes) {
            this.responseSizeInBytes = responseSizeInBytes;
            return this;
        }

        public Builder setRequestHeaders(ArrayList<String> requestHeaders) {
            this.requestHeaders = requestHeaders;
            return this;
        }

        public Builder setResponseHeaders(ArrayList<String> responseHeaders) {
            this.responseHeaders = responseHeaders;
            return this;
        }

        public RequestEntity create() {
            HeaderViewModel requestHeadersModel = new HeaderViewModel(requestHeaders);
            HeaderViewModel responseHeadersModel = new HeaderViewModel(responseHeaders);

            return new RequestEntity(code, method, startDate, url, responseBody, requestBody, tookTime
                    , responseSizeInBytes, requestHeadersModel, responseHeadersModel);
        }
    }
    //endregion

}

File: zoomx/src/main/java/com/zoomx/zoomx/model/RequestEntity.kt
package com.zoomx.zoomx.model

import android.arch.persistence.room.Entity
import android.arch.persistence.room.PrimaryKey
import android.arch.persistence.room.TypeConverters
import com.zoomx.zoomx.db.converters.HeaderConverter
import java.util.*

/**
 * Created by Ahmed Fathallah on 11/20/2017.
 */

@Entity(tableName = "requests")
class RequestEntity {
    //region Setter and getter
    @PrimaryKey(autoGenerate = true)
    var id: Int = 0
    var code: Int = 0
    var method: String? = null
        get() = if (field != null) field else ""
    var startDate: Date? = null
    var url: String? = null
        get() = if (field != null) field else "missing"
    var responseBody: String? = null
        get() = if (field != null) field else ""
    var requestBody: String? = null
        get() = if (field != null) field else ""
    var tookTime: Long = 0
    var responseSizeInBytes: Long = 0

    @TypeConverters(HeaderConverter::class)
    var requestHeaders: HeaderViewModel? = null
    @TypeConverters(HeaderConverter::class)
    var responseHeaders: HeaderViewModel? = null

    constructor() {}

    private constructor(code: Int, method: String?, startDate: Date?, url: String?, responseBody: String?, requestBody: String?, tookTime: Long, responseSizeInBytes: Long, requestHeaders: HeaderViewModel, responseHeaders: HeaderViewModel) {
        this.code = code
        this.method = method
        this.startDate = startDate
        this.url = url
        this.responseBody = responseBody
        this.requestBody = requestBody
        this.tookTime = tookTime
        this.responseSizeInBytes = responseSizeInBytes
        this.requestHeaders = requestHeaders
        this.responseHeaders = responseHeaders
    }

    //endregion

    //region builder
    class Builder {

        private var code: Int = 0
        private var method: String? = null
        private var startDate: Date? = null
        private var url: String? = null
        private var responseBody: String? = null
        private var requestBody: String? = null
        private var tookTime: Long = 0
        private var responseSizeInBytes: Long = 0
        private var requestHeaders: MutableMap<String, String>? = null
        private var responseHeaders: MutableMap<String, String>? = null

        fun setCode(code: Int): Builder {
            this.code = code
            return this
        }

        fun setMethod(method: String): Builder {
            this.method = method
            return this
        }

        fun setStartDate(startDate: Date): Builder {
            this.startDate = startDate
            return this
        }

        fun setUrl(url: String): Builder {
            this.url = url
            return this
        }

        fun setResponseBody(responseBody: String): Builder {
            this.responseBody = responseBody
            return this
        }

        fun setRequestBody(requestBody: String): Builder {
            this.requestBody = requestBody
            return this
        }

        fun setTookTime(tookTime: Long): Builder {
            this.tookTime = tookTime
            return this
        }

        fun setResponseSizeInBytes(responseSizeInBytes: Long): Builder {
            this.responseSizeInBytes = responseSizeInBytes
            return this
        }

        fun setRequestHeaders(requestHeaders: MutableMap<String, String>): Builder {
            this.requestHeaders = requestHeaders
            return this
        }

        fun setResponseHeaders(responseHeaders: MutableMap<String, String>): Builder {
            this.responseHeaders = responseHeaders
            return this
        }

        fun create(): RequestEntity {
            val requestHeadersModel = HeaderViewModel(requestHeaders!!)
            val responseHeadersModel = HeaderViewModel(responseHeaders!!)

            return RequestEntity(code, method, startDate, url, responseBody, requestBody, tookTime, responseSizeInBytes, requestHeadersModel, responseHeadersModel)
        }
    }
    //endregion

}
